src/map.c

```diff
@@ -3,752 +3,630 @@
-     *
-     * If the parameter passed in is NULL, make the element blank instead of
-     * a segfault.
-     */
-    if (!key)
-        memset(node->key, 0, ksize);
-    else
-        memcpy(node->key, key, ksize);
-
-    if (!value)
-        memset(node->data, 0, vsize);
-    else
-        memcpy(node->data, value, vsize);
-
-    return node;
+    node->link.right_red = (map_node_t *) (((size_t) node->link.right_red) | 1);
 }
 
-static void map_delete_node(map_t obj UNUSED, map_node_t *node)
+static inline void rbtn_black_set(map_node_t *node)
 {
-    free(node->key);
-    free(node->data);
-    free(node);
+    node->link.right_red =
+        (map_node_t *) (((size_t) node->link.right_red) & ~3);
 }
 
-/*
- * Perform left rotation with "node". The following happens (with respect
- * to "C"):
- *
- *         B                C
- *        / \              / \
- *       A   C     =>     B   D
- *            \          /
- *             D        A
- *
- * Returns the new node pointing in the spot of the original node.
- */
-static map_node_t *map_rotate_left(map_t obj, map_node_t *node)
+/* Node initializer */
+static inline void rbt_node_init(map_node_t *node)
 {
-    map_node_t *r = node->right, *rl = r->left, *parent = rb_parent(node);
-
-    /* Adjust */
-    rb_set_parent(r, parent);
-    r->left = node;
-
-    node->right = rl;
-    rb_set_parent(node, r);
-
-    if (node->right)
-        rb_set_parent(node->right, node);
-
-    if (parent) {
-        if (parent->right == node)
-            parent->right = r;
-        else
-            parent->left = r;
-    }
-
-    if (node == obj->head)
-        obj->head = r;
-
-    return r;
+    assert((((size_t) node) & (0x1)) == 0);
+    rbtn_left_set(node, NULL);
+    rbtn_right_set(node, NULL);
+    rbtn_red_set(node);
 }
 
-/*
- * Perform a right rotation with "node". The following happens (with respect
- * to "C"):
- *
- *         C                B
- *        / \              / \
- *       B   D     =>     A   C
- *      /                      \
- *     A                        D
- *
- * Return the new node pointing in the spot of the original node.
- */
-static map_node_t *map_rotate_right(map_t obj, map_node_t *node)
+/* Internal utility macros */
+#define rbtn_rotate_left(x_type, x_field, x_node, r_node)  \
+    do {                                                   \
+        (r_node) = rbtn_right_get((x_node));               \
+        rbtn_right_set((x_node), rbtn_left_get((r_node))); \
+        rbtn_left_set((r_node), (x_node));                 \
+    } while (0)
+
+#define rbtn_rotate_right(x_type, x_field, x_node, r_node) \
+    do {                                                   \
+        (r_node) = rbtn_left_get((x_node));                \
+        rbtn_left_set((x_node), rbtn_right_get((r_node))); \
+        rbtn_right_set((r_node), (x_node));                \
+    } while (0)
+
+typedef struct {
+    map_node_t *node;
+    int cmp;
+} internal_map_path_entry_t;
```

Comment:
Think of possible micro-optimizations: encode the value of `cmd` into the following bit representations:
* greater : $11_2$
* equal : $00_2$
* less :  $01_2$

Then, we can even embed the bits of encoded `cmp` into `node` because of ABI alignment. Please evaluate the feasibility and see if we can benefit from such micro-optimization.

Reply:
The prefix `internal_` is a bit annoying. Avoid it if possible.

Reply:
If we can handle enum in compare properly,  we can use the technique from the following example.
Still working it out.
```
#include <stdio.h>

enum Comparison {
    _CMP_LESS = -1,
    _CMP_EQUAL = 0,
    _CMP_GREATER = 1
};

struct Data {
    unsigned int value : 2;
};

int main() {
    struct Data data;

    // Assigning a value to the bit field
    enum Comparison cmp = _CMP_LESS;
    data.value = cmp & 0x3;

    // Accessing and printing the value of the bit field
    printf("Value: %d\n", (int)data.value);

    return 0;
}

```

Reply:
Don't declare `enum Comparison` because `map.h` already provides.